src/node_locks.cc

```diff
@@ -0,0 +1,738 @@
+          for (auto lock_iter = held_locks_iter->second.begin();
+               lock_iter != held_locks_iter->second.end();) {
+            if ((*lock_iter)->env() == env) {
+              lock_iter = held_locks_iter->second.erase(lock_iter);
+            } else {
+              ++lock_iter;
+            }
+          }
+
+          if (held_locks_iter->second.empty()) {
+            held_locks_.erase(held_locks_iter);
+          }
+        }
+      }
+
+      // Wake other environments
+      for (Environment* target_env : envs_to_notify) {
+        if (target_env != env) {
+          WakeEnvironment(target_env);
+        }
+      }
+    }
+
+    // Create and store the new granted lock
+    auto granted_lock =
+        std::make_shared<Lock>(env,
+                               grantable_request->name(),
+                               grantable_request->mode(),
+                               grantable_request->client_id(),
+                               grantable_request->waiting_promise(),
+                               grantable_request->released_promise());
+    {
+      Mutex::ScopedLock scoped_lock(mutex_);
+      held_locks_[grantable_request->name()].push_back(granted_lock);
+    }
+
+    // Call user callback
+    Local<Object> lock_info_obj =
+        CreateLockInfoObject(isolate,
+                             context,
+                             grantable_request->name(),
+                             grantable_request->mode(),
+                             grantable_request->client_id());
+    if (lock_info_obj.IsEmpty()) {
+      return;
+    }
+    Local<Value> callback_arg = lock_info_obj;
+    Local<Value> callback_result;
+    {
+      TryCatchScope try_catch_scope(env);
+      if (!grantable_request->callback()
+               ->Call(context, Undefined(isolate), 1, &callback_arg)
+               .ToLocal(&callback_result)) {
+        grantable_request->waiting_promise()
+            ->Reject(context, try_catch_scope.Exception())
+            .Check();
+        grantable_request->released_promise()
+            ->Reject(context, try_catch_scope.Exception())
+            .Check();
+        return;
+      }
+    }
+
+    // Allocate a LockHolder on the heap to safely manage the lock's lifetime
+    // until the user's callback promise settles.
+    auto* fulfill_holder = new LockHolder(granted_lock);
+    auto* reject_holder = new LockHolder(granted_lock);
+    Local<Function> on_fulfilled_callback;
+    Local<Function> on_rejected_callback;
+
+    if (!Function::New(context,
+                       OnLockCallbackFulfilled,
+                       External::New(isolate, fulfill_holder))
+             .ToLocal(&on_fulfilled_callback) ||
+        !Function::New(context,
+                       OnLockCallbackRejected,
+                       External::New(isolate, reject_holder))
+             .ToLocal(&on_rejected_callback)) {
+      delete fulfill_holder;
+      delete reject_holder;
```

Comment:
These might need to be handled separately. If the first call succeeds, then the first created function will take ownership over the `fulfill_holder`. If the second call then fails for whatever reason, we are deleting the `fulfill_holder` while it's external is still holding the reference to it that it assumes it owns. It would be best to separate this into two separate calls rather than aggregating them together like this. Create one, create it's external and it's function, then create the second...